lib/posix/options/pipe.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (c) 2024, Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdalign.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/sys/fdtable.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/posix/fcntl.h>
+#include <zephyr/zvfs/eventfd.h>
+#include <zephyr/sys/bitarray.h>
+#include <zephyr/sys/fdtable.h>
+
+struct pipe_obj {
+	struct k_poll_signal read_sig;
+	struct k_poll_signal write_sig;
+	struct k_spinlock lock;
+	struct ring_buf buf;
+	int flags;
+};
+
+typedef int (*pipe_rw_op_t)(struct pipe_obj *p, uint8_t *data, size_t size);
+
+static ssize_t pipe_rw_op(void *obj, void *buf, size_t size, pipe_rw_op_t op);
+
+K_MEM_SLAB_DEFINE_STATIC(pipe_slab, CONFIG_POSIX_PIPE_BUF, CONFIG_POSIX_PIPE_MAX,
+			 alignof(uint32_t));
+static struct pipe_obj pipes[CONFIG_POSIX_PIPE_MAX];
+static const struct fd_op_vtable pipe_read_vtable;
+static const struct fd_op_vtable pipe_write_vtable;
+
+static inline bool pipe_is_blocking(struct pipe_obj *p)
+{
+	return (p->flags & O_NONBLOCK) == 0;
+}
+
+static int pipe_poll_prepare(struct pipe_obj *p, struct zsock_pollfd *pfd,
+			     struct k_poll_event **pev, struct k_poll_event *pev_end)
+{
+	return zvfs_poll_prepare(pfd, pev, pev_end, &p->read_sig, &p->write_sig);
+}
+
+static int pipe_poll_update(struct pipe_obj *p, struct zsock_pollfd *pfd, struct k_poll_event **pev)
+{
+	return zvfs_poll_update(pfd, pev, ring_buf_size_get(&p->buf) > 0,
+				ring_buf_space_get(&p->buf) > 0);
+}
+
+static int pipe_read_locked(struct pipe_obj *p, uint8_t *data, size_t size)
+{
+	size_t avail = ring_buf_size_get(&p->buf);
+
+	if (avail == 0) {
+		/* would block / try again */
+		return -EAGAIN;
+	}
+
+	size = MIN(size, avail);
+	ring_buf_get(&p->buf, data, size);
+	avail = ring_buf_size_get(&p->buf);
+
+	if (avail == 0) {
+		k_poll_signal_reset(&p->read_sig);
+	}
+
+	k_poll_signal_raise(&p->write_sig, 0);
+
+	return size;
+}
+
+static int pipe_write_locked(struct pipe_obj *p, uint8_t *data, size_t size)
+{
+	size_t avail = ring_buf_space_get(&p->buf);
+
+	size = MIN(size, avail);
+	ring_buf_put(&p->buf, data, size);
+	avail = ring_buf_space_get(&p->buf);
+
+	if (avail == 0) {
+		k_poll_signal_reset(&p->write_sig);
+	}
+
+	k_poll_signal_raise(&p->read_sig, 0);
+
+	return size;
+}
+
+static ssize_t pipe_read_op(void *obj, void *buf, size_t size)
+{
+	pipe_rw_op_t op = pipe_read_locked;
+
+	return pipe_rw_op(obj, buf, size, op);
+}
+
+static ssize_t pipe_write_op(void *obj, const void *buf, size_t size)
+{
+	pipe_rw_op_t op = pipe_write_locked;
+
+	return pipe_rw_op(obj, (void *)buf, size, op);
+}
+
+static int pipe_close_op(void *obj)
+{
+	struct k_mutex *lock = NULL;
+	struct k_condvar *cond = NULL;
+	struct pipe_obj *p = (struct pipe_obj *)obj;
```

Comment:
Close on the write side should EOF on the read side, close on the read side should generate SIGPIPE on the write side?